src/lib/core/diag.c

```diff
@@ -158,17 +171,43 @@ error_format_msg(struct error *e, const char *format, ...)
 void
 error_append_msg(struct error *e, const char *format, ...)
 {
-	va_list ap;
-	va_start(ap, format);
-	int prefix_len = strlen(e->errmsg);
-	char *msg = e->errmsg + prefix_len;
-	vsnprintf(msg, sizeof(e->errmsg) - prefix_len, format, ap);
-	va_end(ap);
+	va_list ap1, ap2;
+	va_start(ap1, format);
+	va_copy(ap2, ap1);
+	int static_buf_size = sizeof(e->errmsg_buf);
+	int curr_len = strlen(e->errmsg);
+	int len = vsnprintf(NULL, 0, format, ap1);
+	int new_size = curr_len + len + 1;
+	va_end(ap1);
+
+	bool is_malloced = e->errmsg != e->errmsg_buf;
+	if (is_malloced || new_size > static_buf_size) {
+		char *new_buf = xmalloc(new_size);
```

Comment:
@kostja has a good point: we'd better log the error rather than just panic and lose it completely.

Reply:
The original static buffer was there to make sure we can squeak something meaningful even if running out of memory and can't allocate the error message.

Reply:
It's unlikely that `OutOfMemory` error message will be longer than 512 bytes, so it will be printed to the static buffer.

Reply:
> It's unlikely that `OutOfMemory` error message will be longer than 512 bytes, so it will be printed to the static buffer.

The point is that we can run out of memory by coincidence.

Reply:
IMHO the chances that we can allocate error but then fail to allocate message buffer is low.

Anyway OOM is not happened in practice on Linux. See #https://github.com/tarantool/tarantool/issues/3534

Reply:
Okay, I guess if we already have tarantool/tarantool@5a031fb6, then we should follow the current policy.

Reply:
/cc @sergepetrenko

Reply:
> Okay, I guess if we already have https://github.com/tarantool/tarantool/commit/5a031fb6, then we should follow the current policy.

Sounds good.